internal/config/autosave.go

```diff
@@ -1,37 +1,46 @@
 package config
 
 import (
-	"sync"
 	"time"
 )
 
 var Autosave chan bool
-var autotime float64
-
-// lock for autosave
-var autolock sync.Mutex
+var autotime chan float64
 
 func init() {
 	Autosave = make(chan bool)
+	autotime = make(chan float64)
 }
 
 func SetAutoTime(a float64) {
-	autolock.Lock()
-	autotime = a
-	autolock.Unlock()
+	autotime <- a
 }
 
 func StartAutoSave() {
 	go func() {
+		var a float64
+		var t *time.Timer
+		var elapsed <-chan time.Time
 		for {
-			autolock.Lock()
-			a := autotime
-			autolock.Unlock()
-			if a <= 0 {
-				break
+			select {
+			case a = <-autotime:
+				if t != nil && !t.Stop() && len(elapsed) > 0 {
```

Comment:
Why do we check `elapsed` only if `t.Stop()` returned false?

Reply:
According to the docs...

> Stop prevents the [Timer](https://pkg.go.dev/time#Timer) from firing. It returns true if the call stops the timer, false if the timer has already expired or been stopped.

...so in case it returns `true` it shouldn't lead to any elapsed timer yet. For sanity we could decouple it from the return of `Stop()` and just clear `elapsed` in case it isn't empty.

Reply:
> ...so in case it returns `true` it shouldn't lead to any elapsed timer yet.

Aha, I see. If `t.Stop()` returned true, it is guaranteed that the timer has not fired, so the channel is guaranteed to be empty.

BTW I've noticed the doc also says this:

> For a chan-based timer created with NewTimer(d), as of Go 1.23, any receive from t.C after Stop has returned is guaranteed to block rather than receive a stale time value from before the Stop; if the program has not received from t.C already and the timer is running, Stop is guaranteed to return true. **Before Go 1.23, the only safe way to use Stop was insert an extra <-t.C if Stop returned false to drain a potential stale value.**

BTW, I see we are using 1.19, maybe it is a bit too old? (1.23, on the other hand, could be a bit too new, it was only released a week ago.)

> ~in case it isn't empty~ till it is empty.

For the record, it seems guaranteed that `elapsed` will not contain more than 1 element (i.e. there is no difference between `if len(elapsed) > 0` and `for len(elapsed) > 0` in this case)?

Reply:
> BTW, I see we are using 1.19, maybe it is a bit too old? (1.23, on the other hand, could be a bit too new, it was only released a week ago.)

Yes, 1.23 or even 1.22 could lead to problems in the moment deprecated functions have been removed from the standard library.

> For the record, it seems guaranteed that `elapsed` will not contain more than 1 element (i.e. there is no difference between `if len(elapsed) > 0` and `for len(elapsed) > 0` in this case)?

Ok, but since it needs to be checked anyway it doesn't hurt to do it in a loop.